BasicAuthenticationTests#basicAuthenticationWebdriverBidi fails for real on remote Grid jobs since PR #3744

## Summary

Found while attributing nightly `E2E Tests` failures for #3985 (real run data, not the ticket text). `testPackage.legacy.BasicAuthenticationTests#basicAuthenticationWebdriverBidi` (`shaft-engine/src/test/java/testPackage/legacy/BasicAuthenticationTests.java:33-38`) fails for real against `Ubuntu_Chrome_Grid`, with a genuine login failure, not a scope leak and not an OOM:

```
❌ Assertion Failed at Line: 38
🔍 Details: Assertion failed; expected Login Success, but found Login Failure
```

Confirmed stable across the 3 most recent nightly `E2E Tests` runs:

- [30143487701](https://github.com/ShaftHQ/SHAFT_ENGINE/actions/runs/30143487701) (2026-07-25) — `Ubuntu_Chrome_Grid`
- [30066165998](https://github.com/ShaftHQ/SHAFT_ENGINE/actions/runs/30066165998) (2026-07-24) — `Ubuntu_Chrome_Grid`
- [29979003187](https://github.com/ShaftHQ/SHAFT_ENGINE/actions/runs/29979003187) (2026-07-23) — `Ubuntu_Chrome_Grid`

Same error message, same line, all three nights.

## Why this isn't #4006 or the heap-cap issue

- Not a Playwright class (`testPackage.legacy`, not `testPackage.playwright`), so #4006's exclusion doesn't touch it.
- The failure is `java.lang.AssertionError: ❌ Assertion Failed at Line: 38`, not `OutOfMemoryError` — unrelated to the reverted `-Xmx1024m` cap (#4009/#3985).

## Root cause candidate

`navigateToURLWithBasicAuthentication` (`shaft-engine/src/main/java/com/shaft/gui/browser/BrowserActions.java:493-506`) picks a `NATIVE_HAS_AUTHENTICATION` strategy for remote+BiDi execution and does:

```java
((HasAuthentication) driverFactoryHelper.getDriver()).register(uriPredicate, UsernameAndPassword.of(username, password));
```

directly against the driver, with no `Augmenter`. This code path was introduced by PR #3744 (merged 2026-07-18T17:49Z), which closed #3732 ("Remote execution silently downgrades BiDi: basic-auth falls back to URL-embedded credentials off-local"). #3732's own task description called for using Selenium's `Augmenter`/BiDi-augmented `RemoteWebDriver` specifically so `HasAuthentication` would work remotely — that requirement doesn't look implemented in the merged fix.

Before PR #3744, this test always silently downgraded to URL-embedded credentials and therefore always passed (vacuously, without exercising BiDi at all — that was #3732's whole complaint). Since the PR landed, it now actually attempts native remote `HasAuthentication`, and that path appears to not work: the login genuinely fails against `Ubuntu_Chrome_Grid`'s remote Selenium Grid.

Corroborating evidence in the same file: `BasicAuthenticationTests.java` has a *third*, already-disabled test:

```java
@Test(enabled = false, description = "Temporarily disabled: known Selenium BiDi/RemoteWebDriver basic auth instability. Re-enable once upstream issue is resolved.")
public void testBasicAuth() throws MalformedURLException {
```

i.e. BiDi + `RemoteWebDriver` basic-auth instability was already a known, previously-worked-around problem in this exact test class before #3744 changed `basicAuthenticationWebdriverBidi` to exercise the same combination for real.

I could not confirm the exact date this specific test started failing for real (GitHub retains job logs for a limited window; `Ubuntu_Chrome_Grid` logs for the 2026-07-19/07-20 nightly runs, the first ones after PR #3744, have already expired — `HTTP 410`). The 3 most recent nights are the full extent of directly-observable evidence.

## Scope

Read-only diagnosis pass while triaging #3985 nightly failures — no code changes made. Filed per the owner's standing rule that CI failures get reported, not silently absorbed into an unrelated fix.

## Suggested next step

Decide whether `navigateToURLWithBasicAuthentication`'s `NATIVE_HAS_AUTHENTICATION` path needs `Augmenter`-based `RemoteWebDriver` augmentation to make `HasAuthentication` actually intercept remotely (per #3732's original ask), or whether remote BiDi basic-auth should be marked as a known limitation (matching the already-disabled `testBasicAuth` note) and `basicAuthenticationWebdriverBidi` scoped to local-only execution.

--- a/shaft-engine/src/main/java/com/shaft/driver/internal/DriverFactory/DriverFactoryHelper.java
+++ b/shaft-engine/src/main/java/com/shaft/driver/internal/DriverFactory/DriverFactoryHelper.java
@@ -720,7 +720,16 @@ public class DriverFactoryHelper {
                 augmenter = augmenter.addDriverAugmentation(new AddHasAuthentication());
             }
         }
-        return augmenter.augment(driver);
+        var augmentedDriver = augmenter.augment(driver);
+        // Diagnostic evidence for issue #4037: AddHasAuthentication's own register() handler is a silent
+        // no-op unless the augmented driver also implements HasDevTools (CDP-backed), which Selenium only
+        // grants when the session capabilities returned by the remote endpoint advertise a reachable CDP
+        // endpoint (se:cdp/se:cdpEnabled or an equivalent reported URI). Log both so a remote-only failure
+        // can be diagnosed from the returned capabilities instead of guessed at.
+        ReportManagerHelper.logDiscrete("Remote session capabilities after augmentation: `" + driver.getCapabilities()
+                + "`. HasAuthentication=" + (augmentedDriver instanceof HasAuthentication)
+                + ", HasDevTools=" + (augmentedDriver instanceof HasDevTools) + ".", Level.DEBUG);
+        return augmentedDriver;
     }
 
     private static boolean isWindowsPlatform() {

--- a/shaft-engine/src/main/java/com/shaft/gui/browser/BrowserActions.java
+++ b/shaft-engine/src/main/java/com/shaft/gui/browser/BrowserActions.java
@@ -32,6 +32,7 @@ import io.appium.java_client.ios.IOSDriver;
 import io.qameta.allure.Step;
 import org.apache.logging.log4j.Level;
 import org.openqa.selenium.*;
+import org.openqa.selenium.devtools.HasDevTools;
 import org.openqa.selenium.remote.http.HttpRequest;
 import org.openqa.selenium.remote.http.HttpResponse;
 
@@ -493,9 +494,26 @@ public class BrowserActions extends FluentWebDriverAction implements com.shaft.g
             String domainName = browserActionsHelper.getDomainNameFromUrl(targetUrl);
             var strategy = browserActionsHelper.determineBasicAuthenticationStrategy(
                     SHAFT.Properties.platform.executionAddress(), SHAFT.Properties.platform.enableBiDi());
-            if (strategy == BrowserActionsHelper.BasicAuthenticationStrategy.NATIVE_HAS_AUTHENTICATION) {
+            var activeDriver = driverFactoryHelper.getDriver();
+            // HasAuthentication is CDP-backed (Selenium's AddHasAuthentication#getImplementation only
+            // installs the handler when the wrapped driver also implements HasDevTools); Augmenter still
+            // weaves the HasAuthentication interface onto a remote session regardless of CDP reachability
+            // (AddHasAuthentication#isApplicable only checks the browser name), so register() can be
+            // called without throwing yet silently do nothing on a Grid session with no reachable CDP
+            // endpoint. Local drivers (ChromeDriver/EdgeDriver) always implement HasDevTools natively, so
+            // this only gates the remote, Augmenter-based path (issue #4037).
+            boolean isLocalExecution = "local".equalsIgnoreCase(SHAFT.Properties.platform.executionAddress());
+            boolean nativeHandlerCanActuallyWork = activeDriver instanceof HasAuthentication
+                    && (isLocalExecution || activeDriver instanceof HasDevTools);
+            if (strategy == BrowserActionsHelper.BasicAuthenticationStrategy.NATIVE_HAS_AUTHENTICATION && nativeHandlerCanActuallyWork) {
                 Predicate<URI> uriPredicate = uri -> uri.getHost().contains(domainName);
-                ((HasAuthentication) driverFactoryHelper.getDriver()).register(uriPredicate, UsernameAndPassword.of(username, password));
+                ((HasAuthentication) activeDriver).register(uriPredicate, UsernameAndPassword.of(username, password));
+            } else if (strategy == BrowserActionsHelper.BasicAuthenticationStrategy.NATIVE_HAS_AUTHENTICATION) {
+                ReportManagerHelper.logDiscrete("Basic authentication downgraded to URL-embedded credentials: the native "
+                        + "HasAuthentication handler is CDP-backed and this remote driver session does not expose a "
+                        + "reachable Chrome DevTools Protocol endpoint (no HasDevTools augmentation), so register() "
+                        + "would silently do nothing. Falling back to URL-embedded credentials.", Level.WARN);
+                targetUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
             } else {
                 ReportManagerHelper.logDiscrete("Basic authentication downgraded to URL-embedded credentials: execution is remote and "
                         + "SHAFT.Properties.platform.enableBiDi() is disabled, so no native HasAuthentication handler can be provisioned "

--- a/shaft-engine/src/main/java/com/shaft/gui/browser/BrowserActions.java
+++ b/shaft-engine/src/main/java/com/shaft/gui/browser/BrowserActions.java
@@ -490,6 +490,7 @@ public class BrowserActions extends FluentWebDriverAction implements com.shaft.g
     @SuppressWarnings("UnusedReturnValue")
     @Override
     public BrowserActions navigateToURLWithBasicAuthentication(String targetUrl, String username, String password, String targetUrlAfterAuthentication) {
+        String resolvedUrl = targetUrl;
         try {
             String domainName = browserActionsHelper.getDomainNameFromUrl(targetUrl);
             var strategy = browserActionsHelper.determineBasicAuthenticationStrategy(
@@ -513,18 +514,18 @@ public class BrowserActions extends FluentWebDriverAction implements com.shaft.g
                         + "HasAuthentication handler is CDP-backed and this remote driver session does not expose a "
                         + "reachable Chrome DevTools Protocol endpoint (no HasDevTools augmentation), so register() "
                         + "would silently do nothing. Falling back to URL-embedded credentials.", Level.WARN);
-                targetUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
+                resolvedUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
             } else {
                 ReportManagerHelper.logDiscrete("Basic authentication downgraded to URL-embedded credentials: execution is remote and "
                         + "SHAFT.Properties.platform.enableBiDi() is disabled, so no native HasAuthentication handler can be provisioned "
                         + "on the remote driver. Enable BiDi to use the native handler for remote executions.", Level.WARN);
-                targetUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
+                resolvedUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
             }
         } catch (Exception e) {
             ReportManagerHelper.logDiscrete(e);
-            targetUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
+            resolvedUrl = browserActionsHelper.formatUrlForBasicAuthentication(username, password, targetUrl);
         }
-        return navigateToURL(targetUrl, targetUrlAfterAuthentication);
+        return navigateToURL(resolvedUrl, targetUrlAfterAuthentication);
     }
 
     /**

--- a/shaft-engine/src/test/java/testPackage/unitTests/BrowserActionsCoverageUnitTest.java
+++ b/shaft-engine/src/test/java/testPackage/unitTests/BrowserActionsCoverageUnitTest.java
@@ -255,11 +255,45 @@ public class BrowserActionsCoverageUnitTest {
     }
 
     @Test
-    public void navigateWithBasicAuthenticationShouldUseNativeHandlerForRemoteExecutionWhenBiDiIsEnabled() {
-        // Regression for issue #3732: remote execution with BiDi enabled (the default) must use the
-        // real HasAuthentication.register() handler instead of silently downgrading to URL-embedded
-        // credentials, so BasicAuthenticationTests#basicAuthenticationWebdriverBidi actually exercises
-        // the BiDi-augmented path remotely.
+    public void navigateWithBasicAuthenticationShouldUseNativeHandlerForRemoteExecutionWhenBiDiIsEnabledAndCdpIsAvailable() {
+        // Regression for issue #3732: remote execution with BiDi enabled (the default) AND a reachable
+        // CDP endpoint (HasDevTools) must use the real HasAuthentication.register() handler instead of
+        // silently downgrading to URL-embedded credentials, so
+        // BasicAuthenticationTests#basicAuthenticationWebdriverBidi actually exercises the
+        // BiDi-augmented path remotely. See #4037 for the sibling "no CDP access" fallback case.
+        SHAFT.Properties.platform.set().executionAddress("remote-grid");
+        SHAFT.Properties.platform.set().enableBiDi(true);
+
+        WebDriver cdpCapableDriver = mock(WebDriver.class, Mockito.withSettings()
+                .extraInterfaces(JavascriptExecutor.class, TakesScreenshot.class, HasAuthentication.class, HasDevTools.class));
+        WebDriver.Options cdpOptions = mock(WebDriver.Options.class);
+        WebDriver.Navigation cdpNavigation = mock(WebDriver.Navigation.class);
+        when(cdpCapableDriver.manage()).thenReturn(cdpOptions);
+        when(cdpCapableDriver.navigate()).thenReturn(cdpNavigation);
+        when(cdpCapableDriver.getCurrentUrl()).thenReturn("https://example.com/start");
+        doNothing().when(cdpNavigation).to(anyString());
+
+        BrowserActions cdpBrowserActions = new BrowserActions(cdpCapableDriver, true);
+        cdpBrowserActions.navigateToURLWithBasicAuthentication(
+                "https://example.com/secure",
+                "user",
+                "pass",
+                "https://example.com/secure");
+
+        Mockito.verify((HasAuthentication) cdpCapableDriver).register(any(), any());
+        Mockito.verify(cdpNavigation).to("https://example.com/secure");
+    }
+
+    @Test
+    public void navigateWithBasicAuthenticationShouldFallBackWhenRemoteDriverHasNoCdpAccess() {
+        // Regression for issue #4037: Selenium's HasAuthentication is CDP-backed -- its
+        // AddHasAuthentication augmentation only actually installs an auth handler when the wrapped
+        // driver also implements HasDevTools (see AddHasAuthentication#getImplementation). A remote
+        // Augmented session that lacks a reachable CDP endpoint still gets the HasAuthentication
+        // interface woven on by Augmenter#addDependentAugmentations, so register() can be called
+        // without throwing -- but Selenium's handler silently does nothing without HasDevTools, and
+        // the auth prompt is never answered (nightly Ubuntu_Chrome_Grid failure). Detect that case
+        // and use the URL-embedded fallback instead of trusting a handler that cannot actually work.
         SHAFT.Properties.platform.set().executionAddress("remote-grid");
         SHAFT.Properties.platform.set().enableBiDi(true);
         when(driver.getCurrentUrl()).thenReturn("https://example.com/start");
@@ -270,8 +304,8 @@ public class BrowserActionsCoverageUnitTest {
                 "pass",
                 "https://example.com/secure");
 
-        Mockito.verify((HasAuthentication) driver).register(any(), any());
-        Mockito.verify(navigation).to("https://example.com/secure");
+        Mockito.verify((HasAuthentication) driver, Mockito.never()).register(any(), any());
+        Mockito.verify(navigation).to("https://user:pass@example.com/secure");
     }
 
     @Test

